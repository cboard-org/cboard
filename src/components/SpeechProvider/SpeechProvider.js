import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import speech from '../../speech';
import { changeLocale } from '../LanguageProvider/LanguageProvider.actions';

export class SpeechProvider extends Component {
  static propTypes = {
    locale: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired
  };

  componentWillMount() {
    this.setDefaultLocale();
  }

  setDefaultLocale() {
    const { locale, changeLocale } = this.props;

    speech.getLocales().then(speechLocales => {
      if (!locale) {
        const navigatorLocale = navigator.languages[0].slice(0, 2);
        let defaultLocale = speechLocales.includes(navigatorLocale)
          ? navigatorLocale
          : 'en';

        changeLocale(defaultLocale);
      }
    });
  }

  render() {
    return React.Children.only(this.props.children);
  }
}

const mapStateToProps = state => {
  return {
    locale: state.language.locale
  };
};

const mapDispatchToProps = dispatch => ({
  changeLocale: locale => {
    dispatch(changeLocale(locale));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(SpeechProvider);
