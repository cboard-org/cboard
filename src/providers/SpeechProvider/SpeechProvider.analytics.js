import { trackEvent } from '@redux-beacon/google-analytics-gtag';

import {
  CHANGE_VOICE,
  CHANGE_PITCH,
  CHANGE_RATE
} from './SpeechProvider.constants';

const changeVoice = trackEvent((action, prevState, nextState) => ({
  category: 'Speech',
  action: 'Changed Voice',
  label: action.voiceURI
}));

const changePitch = trackEvent((action, prevState, nextState) => ({
  category: 'Speech',
  action: 'Changed Pitch',
  label: action.pitch
}));

const changeRate = trackEvent((action, prevState, nextState) => ({
  category: 'Speech',
  action: 'Changed Rate',
  label: action.rate
}));

const eventsMap = {
  [CHANGE_VOICE]: changeVoice,
  [CHANGE_PITCH]: changePitch,
  [CHANGE_RATE]: changeRate
};

export default eventsMap;
