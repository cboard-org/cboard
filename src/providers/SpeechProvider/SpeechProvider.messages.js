import { defineMessages } from 'react-intl';

export default defineMessages({
  male: {
    id: 'cboard.components.SpeechProvider.male',
    defaultMessage: 'Male'
  },
  female: {
    id: 'cboard.components.SpeechProvider.female',
    defaultMessage: 'Female'
  }
});
