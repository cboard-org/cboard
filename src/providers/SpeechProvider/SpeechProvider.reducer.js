import {
  RECEIVE_VOICES,
  CHANGE_VOICE,
  CHANGE_PITCH,
  CHANGE_RATE,
  CHANGE_VOLUME,
  START_SPEECH,
  END_SPEECH,
  CANCEL_SPEECH,
  EMPTY_VOICES,
  RECEIVE_TTS_ENGINES,
  RECEIVE_TTS_DEFAULT_ENGINE,
  RECEIVE_TTS_ENGINE
} from './SpeechProvider.constants';
import {
  getVoiceURI,
  normalizeLanguageCode,
  standardizeLanguageCode
} from '../../i18n';
import { CHANGE_LANG } from '../LanguageProvider/LanguageProvider.constants';
import { LOGIN_SUCCESS } from '../../components/Account/Login/Login.constants';
import { DEFAULT_LANG } from '../../components/App/App.constants';

const initialState = {
  voices: [],
  ttsEngines: [],
  ttsDefaultEngine: {},
  ttsEngine: {},
  langs: [],
  options: {
    lang: '',
    voiceURI: null,
    isCloud: null,
    pitch: 1.0,
    rate: 1.0,
    volume: 1
  },
  isSpeaking: false
};

function speechProviderReducer(state = initialState, action) {
  switch (action.type) {
    case LOGIN_SUCCESS:
      const settings = action.payload.settings || {};
      const { speech } = settings || {};

      const pitch = speech && speech.pitch ? speech.pitch : state.options.pitch;
      const rate = speech && speech.rate ? speech.rate : state.options.rate;

      const options = { ...state.options, pitch, rate };

      return {
        ...state,
        options
      };
    case RECEIVE_VOICES:
      const langs = action.voices.map(voice =>
        normalizeLanguageCode(standardizeLanguageCode(voice.lang))
      );
      //hack just for Alfanum Serbian voices
      //https://github.com/cboard-org/cboard/issues/715
      if (langs.includes('sr-RS')) {
        langs.push('sr-SP');
      }
      //hack just for Tetum language
      //https://github.com/cboard-org/cboard/issues/848
      if (langs.includes('pt-BR') || langs.includes('pt-PT')) {
        langs.push('pt-TL');
      }
      return {
        ...state,
        voices: action.voices,
        langs: [...new Set(langs)].sort()
      };
    case CHANGE_VOICE:
      return {
        ...state,
        options: {
          ...state.options,
          voiceURI: action ? action.voiceURI : EMPTY_VOICES,
          lang: action ? action.lang : DEFAULT_LANG,
          isCloud: action ? action.isCloud || null : null
        }
      };
    case RECEIVE_TTS_ENGINES:
      return {
        ...state,
        ttsEngines: action.ttsEngines
      };
    case RECEIVE_TTS_DEFAULT_ENGINE:
      return {
        ...state,
        ttsDefaultEngine: action.ttsDefaultEngine,
        ttsEngine: state.ttsEngine ? state.ttsEngine : action.ttsDefaultEngine
      };
    case RECEIVE_TTS_ENGINE:
      const newTtsEngine = state.ttsEngines.find(
        engine => engine.name === action.ttsEngineName
      );
      return {
        ...state,
        ttsEngine: newTtsEngine ? newTtsEngine : state.ttsEngine
      };
    case CHANGE_LANG:
      //hack just for Alfanum Serbian voices
      //https://github.com/cboard-org/cboard/issues/715
      if (action.lang === 'sr-SP' || action.lang === 'sr-RS') {
        const language = 'sr-RS';
        return {
          ...state,
          options: {
            ...state.options,
            voiceURI:
              state.options.lang !== language
                ? getVoiceURI(language, state.voices)
                : state.options.voiceURI,
            lang: language
          },
          langs: ['sr-SP', 'sr-RS']
        };
      } else {
        return {
          ...state,
          options: {
            ...state.options,
            voiceURI:
              state.options.lang.substring(0, 2) !== action.lang.substring(0, 2)
                ? getVoiceURI(action.lang, state.voices)
                : state.options.voiceURI,
            lang:
              state.options.lang.substring(0, 2) !== action.lang.substring(0, 2)
                ? action.lang
                : state.options.lang
          }
        };
      }
    case CHANGE_PITCH:
      return { ...state, options: { ...state.options, pitch: action.pitch } };
    case CHANGE_RATE:
      return { ...state, options: { ...state.options, rate: action.rate } };
    case CHANGE_VOLUME:
      return { ...state, options: { ...state.options, volume: action.volume } };
    case START_SPEECH:
      return { ...state, isSpeaking: action.isSpeaking };
    case END_SPEECH:
      return { ...state, isSpeaking: action.isSpeaking };
    case CANCEL_SPEECH:
      return { ...state, isSpeaking: action.isSpeaking };
    default:
      return state;
  }
}

export default speechProviderReducer;
