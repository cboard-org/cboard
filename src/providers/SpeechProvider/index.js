export { default } from './SpeechProvider.container';
