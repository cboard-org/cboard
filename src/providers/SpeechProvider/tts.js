import { normalizeLanguageCode } from '../../i18n';

const synth = window.speechSynthesis;
let voices = [];

const tts = {
  isSupported() {
    return 'speechSynthesis' in window;
  },

  normalizeVoices(voices) {
    return voices.map(({ voiceURI, name, lang }) => ({
      voiceURI,
      name,
      lang: normalizeLanguageCode(lang)
    }));
  },

  getVoiceByLang(lang) {
    return this.getVoices().then(voices =>
      voices.find(voice => voice.lang === lang)
    );
  },

  getVoiceByVoiceURI(VoiceURI) {
    return this.getVoices().then(voices =>
      voices.find(voice => voice.voiceURI === VoiceURI)
    );
  },

  getLangs() {
    return this.getVoices().then(voices => {
      const langs = [...new Set(voices.map(voice => voice.lang))];
      return langs;
    });
  },

  getVoices() {
    if (voices.length) {
      return Promise.resolve(voices);
    }

    return new Promise((resolve, reject) => {
      // iOS
      voices = synth.getVoices() || [];
      if (voices.length) {
        resolve(this.normalizeVoices(voices));
      }

      // Android
      if ('onvoiceschanged' in synth) {
        speechSynthesis.onvoiceschanged = () => {
          voices = synth.getVoices();
          resolve(this.normalizeVoices(voices));
        };
      }
    });
  },

  cancel() {
    synth.cancel();
  },

  speak(text, { voiceURI, pitch = 1, rate = 1, volume = 1 }) {
    this.getVoiceByVoiceURI(voiceURI).then(voice => {
      const msg = new SpeechSynthesisUtterance(text);
      msg.voice = voice;
      msg.name = voice.name;
      msg.lang = voice.lang;
      msg.voiceURI = voice.voiceURI;
      msg.pitch = pitch;
      msg.rate = rate;
      msg.volume = volume;
      synth.speak(msg);
    });
  }
};

export default tts;
