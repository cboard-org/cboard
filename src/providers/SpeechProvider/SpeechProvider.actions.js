import {
  REQUEST_VOICES,
  RECEIVE_VOICES,
  CHANGE_VOICE,
  CHANGE_PITCH,
  CHANGE_RATE,
  START_SPEECH,
  END_SPEECH,
  CANCEL_SPEECH
} from './SpeechProvider.constants';

import tts from './tts';

export function requestVoices() {
  return {
    type: REQUEST_VOICES
  };
}

export function receiveVoices(voices) {
  return {
    type: RECEIVE_VOICES,
    voices
  };
}

export function changeVoice(voiceURI, lang) {
  return {
    type: CHANGE_VOICE,
    voiceURI,
    lang
  };
}

export function changePitch(pitch) {
  return {
    type: CHANGE_PITCH,
    pitch
  };
}

export function changeRate(rate) {
  return {
    type: CHANGE_RATE,
    rate
  };
}

export function getVoices() {
  return dispatch => {
    dispatch(requestVoices());

    return tts.getVoices().then(voices => {
      voices = voices.map(({ voiceURI, lang, name }) => ({
        voiceURI,
        lang,
        name
      }));
      dispatch(receiveVoices(voices));
    });
  };
}

function startSpeech() {
  return {
    type: START_SPEECH,
    isSpeaking: true
  };
}

function endSpeech() {
  return {
    type: END_SPEECH,
    isSpeaking: false
  };
}

export function cancelSpeech() {
  return dispatch => {
    dispatch({ type: CANCEL_SPEECH });
    tts.cancel();
  };
}

export function speak(text) {
  return (dispatch, getState) => {
    const options = getState().speech.options;
    dispatch(startSpeech());

    tts.speak(text, {
      ...options,
      onend: event => {
        dispatch(endSpeech());
      }
    });
  };
}
