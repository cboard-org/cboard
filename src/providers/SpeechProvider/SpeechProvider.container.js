import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import tts from './tts';
import {
  getVoices,
  getTtsEngines,
  getTtsDefaultEngine,
  updateLangSpeechStatus,
  setTtsEngine,
  setCurrentVoiceSource
} from './SpeechProvider.actions';
import { isAndroid } from '../../cordova-util';

export class SpeechProvider extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    ttsEngine: PropTypes.object,
    setTtsEngine: PropTypes.func
  };

  async componentDidMount() {
    const {
      getVoices,
      updateLangSpeechStatus,
      getTtsEngines,
      getTtsDefaultEngine,
      ttsEngine,
      setTtsEngine,
      setCurrentVoiceSource
    } = this.props;

    if (tts.isSupported()) {
      //if android we have to set the tts engine first
      if (isAndroid()) {
        getTtsEngines();
        getTtsDefaultEngine();
      }
      if (ttsEngine && ttsEngine.name) {
        try {
          await setTtsEngine(ttsEngine.name);
        } catch (err) {
          console.error(err.message);
        }
      }
      try {
        const voices = await getVoices();
        await updateLangSpeechStatus(voices);
      } catch (err) {
        console.error(err.message);
      }
    }
    setCurrentVoiceSource();
  }

  render() {
    const { children } = this.props;

    return React.Children.only(children);
  }
}

const mapStateToProps = state => ({
  ttsEngine: state.speech.ttsEngine
});

const mapDispatchToProps = {
  getVoices,
  getTtsEngines,
  getTtsDefaultEngine,
  setTtsEngine,
  updateLangSpeechStatus,
  setCurrentVoiceSource
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SpeechProvider);
