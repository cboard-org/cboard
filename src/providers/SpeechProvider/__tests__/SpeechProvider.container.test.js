import { SpeechProvider } from '../SpeechProvider.container';
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

jest.mock('react-redux');

describe('container group Test', () => {
  const props = {
    voices: [],
    langs: [],
    children: {},
    getVoices: jest.fn()
  };
  it('speechProviderContainer: snapshot test', () => {
    const snapshot = renderer
      .create(
        <SpeechProvider {...props}>
          <div>child one</div>
          <div>child two</div>
        </SpeechProvider>
      )
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
