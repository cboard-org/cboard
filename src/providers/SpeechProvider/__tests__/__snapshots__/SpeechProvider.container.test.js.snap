// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container group Test speechProviderContainer: snapshot test 1`] = `null`;
